basil/HL/tektronix_oscilloscope.py

```diff
@@ -0,0 +1,183 @@
+
+            # calculate horizontal scale
+            slope = float(xincr)
+            offset = float(pt_off) * -slope + float(xzero)
+            unit = xunit.strip('"')
+            result = XScale(slope, offset, unit)
+        return result
+
+    def _get_yscale(self, channel=1):
+        scale = float(self._intf.query("{}:SCALE?".format('CH' + str(channel))))
+        position = float(self._intf.query("{}:POSITION?".format('CH' + str(channel))))
+        top = scale * (5 - position)
+        bottom = scale * (-5 - position)
+        return YScale(top=top, bottom=bottom)
+
+    def _has_data_available(self, source):
+        """Checks that the source seems to have actual data available for download.
+        This function should be called before performing a curve query."""
+        # Use "display:global:CH{x}:state?" to determine if the channel is displayed
+        # and available for download
+        if source == "NONE":
+            return False
+        return bool(int(self._intf.query("display:global:{}:state?".format(source))))
+
+    def _setup_curve_query(self, parameters, channel=1):
+        """Setup the instrument for the curve query operation"""
+
+        # extract the job parameters
+        wave_type, channel, encoding, bit_nr, datatype, rec_len = parameters['CH' + str(channel)]
+
+        # Switch to the source and setup the data encoding
+        self._intf.write("data:source {}".format('CH' + channel))
+        self.set_data_encoding('ascii')
+        self._intf.write("WFMOUTPRE:BIT_NR {}".format(bit_nr))
+
+        # Set the start and stop point of the record
+        rec_len = self._intf.query("horizontal:recordlength?").strip()
+        self._intf.write("data:start 1")
+        self._intf.write("data:stop {}".format(rec_len))
+        return wave_type
+
+    def _post_process_analog(self, source_data, x_scale, channel=1):
+        """Post processes analog channel data"""
+
+        # Normal analog channels must have the vertical scale and offset applied
+        offset = float(self._intf.query("WFMOutpre:YZEro?"))
+        scale = float(self._intf.query("WFMOutpre:YMUlt?"))
+        source_data = [scale * i + offset for i in source_data]
+
+        # Include y-scale information with analog channel waveforms
+        y_scale = self._get_yscale(channel=channel)
+
+        return channel, source_data, x_scale, y_scale
+
+    def get_waveform(self, channel=1):
+        """Returns an iterator that yields the source data from the oscilloscope"""
+
+        jobs = self._make_jobs(channel=channel)
+
+        # remember the state of the acquisition system and then stop acquiring waveforms
+        acq_state = self._intf.query("ACQuire:STATE?").strip()
+        self._intf.write("ACQuire:STATE STOP")
+        wave_type = self._setup_curve_query(jobs, channel=channel)
+
+        # Horizontal scale information
+        x_scale = self._get_xscale()
+        if x_scale is not None:
```

Comment:
In case that `x_scale` is None, trying to return `ret_val` will result in a `NameError`. Either declare it in the beginning as `None` or move the return statement inside the `if x_scale is not None` branch